Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1172,33 +1175,73 @@ void RenderBlockFlow::adjustPositionedBlock(RenderBox& child, const MarginInfo&
 
+    // The non-standard WebKitCenter value should be overriden by justify-self.
+    if ((child.style().justifySelf().position() == ItemPosition::Auto && style().textAlign() == TextAlignMode::WebKitCenter) || child.style().marginStartUsing(&style()).isAuto()) {
+        // If the child is being centred then the margin calculated to do that has factored in any offset required to
+        // avoid floats, so use it if necessary.
+        startPosition = std::max(startPosition, positionToAvoidFloats + childMarginStart);
+        setLogicalLeftForChild(child, style().isLeftToRightDirection() ? startPosition : totalAvailableLogicalWidth - startPosition - logicalWidthForChild(child), applyDelta);
+        return;
+    }
+
+    auto computeSelfAlignedLogicalLeftPositionForChild = [&](LayoutUnit currentPosition) -> LayoutUnit {
+        auto startAlignedPosition = currentPosition;
+        auto endAlignedPosition = totalAvailableLogicalWidth - startAlignedPosition - logicalWidthForChild(child);
+        auto selfAlignmentStyle = child.style().resolvedJustifySelf(&style(), ItemPosition::Normal);
+        const auto parentDirection = style().direction();
+        if (!isLeftToRightDirection(parentDirection))
+            std::swap(startAlignedPosition, endAlignedPosition);
+        if (child.style().justifySelf().overflow() == OverflowAlignment::Safe && logicalWidthForChild(child) > totalAvailableLogicalWidth)
+            return startAlignedPosition;
+        const auto childDirection = child.style().direction();
+        if (selfAlignmentStyle.isStartward(childDirection, parentDirection))
+            return startAlignedPosition;
+        if (selfAlignmentStyle.isEndward(childDirection, parentDirection))
+            return endAlignedPosition;
+        if (selfAlignmentStyle.isCentered())
+            return totalAvailableLogicalWidth / 2 - logicalWidthForChild(child) / 2;
+        ASSERT_NOT_REACHED();
+        return { };
+    };
+
     // If the child has an offset from the content edge to avoid floats then use that, otherwise let any negative
     // margin pull it back over the content edge or any positive margin push it out.
-    // If the child is being centred then the margin calculated to do that has factored in any offset required to
-    // avoid floats, so use it if necessary.
-
-    if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto())
-        newPosition = std::max(newPosition, positionToAvoidFloats + childMarginStart);
-    else if (positionToAvoidFloats > initialStartPosition)
-        newPosition = std::max(newPosition, positionToAvoidFloats);
+    if (positionToAvoidFloats > startPositionWithOnlyBorderAndPadding)
+        startPosition = std::max(startPosition, positionToAvoidFloats);
+    setLogicalLeftForChild(child, computeSelfAlignedLogicalLeftPositionForChild(startPosition), applyDelta);
```

Comment:
Do we not need to floor the result of `computeSelfAlignedLogicalLeftPositionForChild()` by `positionToAvoidFloats`?

Do we have a testcase covering this case?

Reply:
I don't think so because if there is a `positionToAvoidFloats`, then the `startAlignedPosition` used within `computeSelfAlignedLogicalLeftPositionForChild` takes that into account. A test case would be good here though and could probably be re-written to make it more clear